### ISSUE TITLE
Workchain cancels without error when dataclass is in context

### ISSUE THREAD

**Issue by janssenhenning** [aiidateam/aiida-core#5775]
- [x] [AiiDA Troubleshooting Documentation](https://aiida.readthedocs.io/projects/aiida-core/en/latest/intro/troubleshooting.html)
- [x] [AiiDA Users Forum](https://groups.google.com/forum/#!forum/aiidausers)

### Describe the bug

The following workchain finishes without calling the results step, but does also not indicate any exception. The only hint
of this is that the output `x` is missing.

```python
from aiida.engine import WorkChain, calcfunction
from aiida import orm
import dataclasses as dc

@dc.dataclass
class Data:
    x: int
    y: int

class DataClassWorkChain(WorkChain):
    
    @classmethod
    def define(cls, spec):
        
        super().define(spec)
        spec.output('x', valid_type=orm.Int)
        spec.outline(cls.run, cls.results)
        
    def run(self):
        self.ctx.data = Data(x=2,y=-4)
        self.report(f'Stored: {self.ctx.data}')
    
    def results(self):
        
        self.report(f'Loaded: {self.ctx.data}')
        self.out('x', get_output(self.ctx.data.x, self.ctx.data.y))

@calcfunction
def get_output(x, y):
    return x + y
```
Finished process
```
Property     Value
-----------  -------------------------------------------------------------
type         DataClassWorkChain
state        Finished [11] The process did not register a required output.
pk           2219
uuid         f3f2a1b3-8d9c-4f58-9fb8-743fec14ad5c
label
description
ctime        2022-11-17 10:33:32.100126+00:00
mtime        2022-11-17 10:33:32.399256+00:00

Log messages
---------------------------------------------
There are 1 log messages for this calculation
Run 'verdi process report 2219' to see them
```
Log messages
```
2022-11-17 10:33:32 [1087 | REPORT]: [2219|DataClassWorkChain|run]: Stored: Data(x=2, y=-4)
```

### Expected behavior

Either the workchain should correctly persist the dataclass or show a clear error message, not fail silently

### Your environment

- Operating system [e.g. Linux]: Linux
- Python version [e.g. 3.7.1]: 3.8.10
- aiida-core version [e.g. 1.2.1]: 2.1.1

### Additional context

Merging the two steps, leads to the workchain running successfully, so propably the checkpointing is at fault somehow.

**Comment by sphuber** [comment]
Indeed, I think that dataclasses are not supported by our serialization. In principle I think it should be possible to add it for a `dataclass` instance, but it depends on the type of attributes. If they are simple YAML-serializable attributes then it is no problem. Might make sense to add support for these cases. If you are interested in having a stab at implementing it, you essentially have to add a hook for `dataclasses.dataclass` in `aiida.orm.utils.serialize`. Let me know if you are interested and/or need any help.

**Comment by sphuber** [comment]
Can you check the daemon log if there is an exception. There should have been an exception logged when `aiida.engine.persistence.AiiDAPersister.save_checkpoint` was called. Not sure why this doesn't show up in the process report.

**Comment by janssenhenning** [comment]
The only log message I see is the `Stored: Data(x=2, y=-4)` one I put into the workchain. I see no exception at all

**Comment by sphuber** [comment]
> The only log message I see is the `Stored: Data(x=2, y=-4)` one I put into the workchain. I see no exception at all

Confirmed, will have a look.

**Comment by sphuber** [comment]
Found the problem :D it's a classic: The `run` method is actually a method on the `Process` base class. That is triggered to run the whole outline, but if you override it, that is all it does as the logic to continue to the next step is in the base class and it is now never called. If you change `run` to anything else, it will work. Seems that serialization of dataclasses is already supported. I tried your example and it works, also when sending over the daemon where checkpointing is enabled.

This has been reported before in this issue: https://github.com/aiidateam/aiida-core/issues/5467 I will look into this soon to finally fix it.
